nomad/state/state_store.go

```diff
@@ -4100,6 +4115,83 @@ type StateSnapshot struct {
 	StateStore
 }
 
+// DenormalizeAllocationsMap takes in a map of nodes to allocations, and queries the
+// Allocation for each of the Allocation diffs and merges the updated attributes with
+// the existing Allocation, and attaches the Job provided
+func (s *StateSnapshot) DenormalizeAllocationsMap(nodeAllocations map[string][]*structs.Allocation, job *structs.Job) error {
+	for nodeID, allocs := range nodeAllocations {
+		denormalizedAllocs, err := s.DenormalizeAllocationSlice(allocs, job)
+		if err != nil {
+			return err
+		}
+
+		nodeAllocations[nodeID] = denormalizedAllocs
+	}
+	return nil
+}
+
+// DenormalizeAllocationSlice queries the Allocation for each allocation diff
+// represented as an Allocation and merges the updated attributes with the existing
+// Allocation, and attaches the Job provided.
+func (s *StateSnapshot) DenormalizeAllocationSlice(allocs []*structs.Allocation, job *structs.Job) ([]*structs.Allocation, error) {
+	allocDiffs := make([]*structs.AllocationDiff, len(allocs))
+	for i, alloc := range allocs {
+		allocDiffs[i] = alloc.AllocationDiff()
+	}
+
+	return s.DenormalizeAllocationDiffSlice(allocDiffs, job)
+}
+
+// DenormalizeAllocationDiffSlice queries the Allocation for each AllocationDiff and merges
+// the updated attributes with the existing Allocation, and attaches the Job provided
+func (s *StateSnapshot) DenormalizeAllocationDiffSlice(allocDiffs []*structs.AllocationDiff, job *structs.Job) ([]*structs.Allocation, error) {
+	// Output index for denormalized Allocations
+	j := 0
+
+	denormalizedAllocs := make([]*structs.Allocation, len(allocDiffs))
+	for _, allocDiff := range allocDiffs {
+		alloc, err := s.AllocByID(nil, allocDiff.ID)
+		if err != nil {
+			return nil, fmt.Errorf("alloc lookup failed: %v", err)
+		}
+		if alloc == nil {
+			continue
```

Comment:
Continuing the discussion from
https://github.com/hashicorp/nomad/pull/5407/files/c242adea8786e7541c8c108c342026730b028ff4#diff-ccbd515c67aa55098b48f1106de134aaR4134

There are a few reasons I don't think returning an error is the right approach here:
1) We do diffs for only stopped or preempted allocs. If they don’t exist, it means they have already been stopped/preempted. If that's the case, it’s fine for the update to not make it to the alloc in that case.
2) The raft log has already been dispatched at this point, and this code just applies the log to the state store. I'm not sure it'd even be 'correct' to not apply this to the state store, given that we can't remove the raft log entry. Wouldn't the state store be stuck whenever it has to apply this log entry?
3) It looks like existing code for preempted allocs dealt with this in the same way. https://github.com/hashicorp/nomad/blob/987ed01b110d59defc2c8135fe9b303b244d8477/nomad/state/state_store.go#L236

Reply:
1) should not happen unless the plan applier is acting on stale state store data.  We asked about returning an error so that its very visible when this happens. ie. this suggestion was for safety 

Re: 2) - returning the error will fail the optimistic apply to the leader's state store in https://github.com/hashicorp/nomad/blob/95297c608c59fc66bb965c9231a0685eb3ddeac1/nomad/plan_apply.go#L226 and propagate up to the worker in https://github.com/hashicorp/nomad/blob/95297c608c59fc66bb965c9231a0685eb3ddeac1/nomad/worker.go#L324, eventually that will fail the evaluation and retry upon which the scheduler can make a new plan with updated information. The state store will always fail on that entry, but a future entry with correct information should succeed.  Also see https://github.com/hashicorp/nomad/blob/95297c608c59fc66bb965c9231a0685eb3ddeac1/nomad/plan_queue.go#L20 and how we block on the future from raft in https://github.com/hashicorp/nomad/blob/95297c608c59fc66bb965c9231a0685eb3ddeac1/nomad/plan_endpoint.go#L43. 

3) that logic could be worth changing as well, I will address that in a future PR since I am going to be touching the plan applier again after this PR is merged, you don't need to change that right now though.